lib/services/local/saml_idp_service_provider.go

```diff
@@ -37,157 +39,97 @@ const (
 
 // SAMLIdPServiceProviderService manages IdP service providers in the Backend.
 type SAMLIdPServiceProviderService struct {
-	backend.Backend
+	svc *generic.Service[types.SAMLIdPServiceProvider]
 }
 
 // NewSAMLIdPServiceProviderService creates a new SAMLIdPServiceProviderService.
 func NewSAMLIdPServiceProviderService(backend backend.Backend) *SAMLIdPServiceProviderService {
-	return &SAMLIdPServiceProviderService{Backend: backend}
-}
-
-// ListSAMLIdPServiceProviders returns a paginated list of SAML IdP service provider resources.
-func (s *SAMLIdPServiceProviderService) ListSAMLIdPServiceProviders(ctx context.Context, pageSize int, pageToken string) ([]types.SAMLIdPServiceProvider, string, error) {
-	rangeStart := backend.Key(samlIDPServiceProviderPrefix, pageToken)
-	rangeEnd := backend.RangeEnd(rangeStart)
-
-	// Adjust page size, so it can't be too large.
-	if pageSize <= 0 || pageSize > samlIDPServiceProviderMaxPageSize {
-		pageSize = samlIDPServiceProviderMaxPageSize
-	}
-
-	// Increment pageSize to allow for the extra item represented by nextKey.
-	// We skip this item in the results below.
-	limit := pageSize + 1
-	var out []types.SAMLIdPServiceProvider
-
-	// no filter provided get the range directly
-	result, err := s.GetRange(ctx, rangeStart, rangeEnd, limit)
+	svc, err := generic.NewService(&generic.ServiceConfig[types.SAMLIdPServiceProvider]{
+		Backend:       backend,
+		PageLimit:     samlIDPServiceProviderMaxPageSize,
+		ResourceKind:  types.KindSAMLIdPServiceProvider,
+		BackendPrefix: samlIDPServiceProviderPrefix,
+		MarshalFunc:   services.MarshalSAMLIdPServiceProvider,
+		UnmarshalFunc: services.UnmarshalSAMLIdPServiceProvider,
+	})
 	if err != nil {
-		return nil, "", trace.Wrap(err)
-	}
-
-	out = make([]types.SAMLIdPServiceProvider, 0, len(result.Items))
-	for _, item := range result.Items {
-		sp, err := services.UnmarshalSAMLIdPServiceProvider(item.Value)
-		if err != nil {
-			return nil, "", trace.Wrap(err)
-		}
-		out = append(out, sp)
+		panic(fmt.Sprintf("SAML IdP service provider local service is not configured properly: %v", err))
```

Comment:
I wonder if simply panicking the error would be good enough here.

I'd also add a `var _ = NewSAMLIdPServiceProviderService((*memory.Backend)(nil))` to the package, just to be sure that the panic is triggered very early.